File: leetcode/106.construct-binary-tree-from-Inorder-and-postorder-traversal/ts/buildtree.ts

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const buildTree = (inorder: number[], postorder: number[]): TreeNode | null => {
    if (inorder.length === 0 || inorder.length !== postorder.length) return null

    const prev = new TreeNode()
    traversal(inorder, 0, inorder.length, postorder, 0, postorder.length, prev, true)
    return prev.left
}

const traversal = (inorder: number[], inLeft: number, inRight: number, postorder: number[], postLeft: number, postRight: number, prev: TreeNode, isLeft: boolean) => {
    if (inLeft >= inRight) return

    const rootVal = postorder[postRight - 1]
    const root = new TreeNode(rootVal)
    if (isLeft) prev.left = root
    else prev.right = root

    const splitIdx = findIdx(inorder, inLeft, inRight, rootVal)
    postRight--

    const leftLen = splitIdx - inLeft

    traversal(inorder, inLeft, splitIdx, postorder, postLeft, postLeft + leftLen, root, true)
    traversal(inorder, splitIdx + 1, inRight, postorder, postLeft + leftLen, postRight, root, false)
}

const findIdx = (nums: number[], left: number, right: number, target: number): number => {
    for (let i = left; i < right; i++)
        if (nums[i] === target) return i
    return -1
}

export default buildTree

File: leetcode/106.construct-binary-tree-from-Inorder-and-postorder-traversal/go/buildtree.go
package constructbinarytreefrominorderandpostordertraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(inorder) == 0 || len(inorder) != len(postorder) {
		return nil
	}

	prev := TreeNode{}
	traversal(inorder, 0, len(inorder), postorder, 0, len(postorder), &prev, false)
	return prev.Right
}

func traversal(inorder []int, inLeft int, inRight int, postorder []int, postLeft int, postRight int, prev *TreeNode, isLeft bool) {
	if inLeft >= inRight {
		return
	}

	rootVal := postorder[postRight-1]
	root := &TreeNode{Val: rootVal}
	if isLeft {
		prev.Left = root
	} else {
		prev.Right = root
	}

	splitIdx := findIdx(inorder, inLeft, inRight, rootVal)
	leftSize := splitIdx - inLeft
	postRight--

	traversal(inorder, inLeft, splitIdx, postorder, postLeft, postLeft+leftSize, root, true)
	traversal(inorder, splitIdx+1, inRight, postorder, postLeft+leftSize, postRight, root, false)
}

func findIdx(nums []int, left int, right int, target int) int {
	for i := left; i < right; i++ {
		if nums[i] == target {
			return i
		}
	}
	return -1
}
